### ISSUE TITLE
NaN from finite Q/K/V on V100 FP16 self-attention (head_dim=128, seq_len=2925)

### ISSUE THREAD

**Issue by buxinzi2233** [ai-bond/flash-attention-v100#43]
Hello,

Thank you for maintaining FlashAttention for Volta. I am testing the public
FA2-compatible API as the attention backend for Anima DiT training on a Tesla
V100. I found a numerical failure where finite FP16 Q/K/V tensors produce a
non-finite dense self-attention output on the first training step.

This case does **not** use ComfyUI, a custom node, a monkey patch, a compatibility
shim, or private `_wrapped_flash_attn_*` functions. The host calls the package's
public `flash_attn_func` directly. It is therefore related to, but materially
different from, #36, where the integration layer remained a possible cause.

## Environment

| Item | Value |
| --- | --- |
| GPU | NVIDIA Tesla V100-SXM2-16GB |
| Compute capability | SM 7.0 (Volta) |
| Python | 3.12 (`cp312` wheel) |
| PyTorch | `2.10.0+cu129` |
| CUDA | runtime 12.9 / toolkit 12.9.1 |
| Package | `flash_attn_v100-26.6-cp312-cp312-linux_x86_64.whl` |
| Package facade | `flash-attn 2.8.3`, backend `v26.06` |
| Wheel SHA-256 | `74b4cdbd3a225745be82ec44db4104a89bb77094119feec1cc1bf99c7bc1a6a2` |
| Precision | FP16 autocast |
| Q/K/V | shape `(1, 2925, 16, 128)`, dtype `torch.float16` |
| Output | raw shape `(1, 2925, 16, 128)`, then flattened to `(1, 2925, 2048)` |

## Exact API call

The model dispatcher passes tensors in the documented
`(batch, seqlen, nheads, headdim)` layout:

```python
from flash_attn import flash_attn_func

out = flash_attn_func(
    q,
    k,
    v,
    dropout_p=0.0,
    softmax_scale=None,  # 1 / sqrt(128)
)
```

This is the dense path. The package wrapper permutes and makes Q/K/V contiguous
before calling `flash_attn_v100_cuda.fwd`.

## Isolation results

| Configuration | Result |
| --- | --- |
| PyTorch SDPA for self- and cross-attention | Multiple forward/backward steps remain finite. |
| `flash_attn_func` for self-attention, Torch SDPA for cross-attention | First-step NaN still occurs in the self-attention output. |
| `flash_attn_func` for both attention kinds with fail-fast checks | Q/K/V pass the pre-dispatch checks; the first non-finite tensor is the returned self-attention output. |
| Torch SDPA with `torch.compile` | Works on this V100. |
| This FlashAttention path with `torch.compile` | Also hits a Dynamo shape-guard/traceability issue; eager execution was used for the numerical comparison. |

The first fail-fast diagnostic is:

```text
self_attn.attention_output
backend=flash
dtype=torch.float16
shape=(1, 2925, 2048)
```

The finite checks run in this order:

1. Q, K, and V immediately before backend dispatch: finite.
2. Attention result immediately after dispatch and layout flattening: non-finite.
3. Output projection, residual update, loss, and backward have not yet run when
   the exception is raised.

The only operations between the CUDA return and the check are layout transpose
and flattening. Replacing the backend call with PyTorch SDPA for the same model
input remains finite.

## Separate FP16 overflow ruled out

Anima also had an independent FP16 residual overflow: a trained AdaLN
`gate * branch` product could exceed 65504 before the residual add. We added an
FP32 product/residual/final-projection guard for that issue. With this guard
enabled, the FlashAttention return value still becomes non-finite at the
boundary above, so this report is not about the residual overflow.

## Questions

1. Is `SM70 + FP16 + head_dim=128 + seqlen=2925` supported by backend `v26.06`?
   In particular, is a non-tile-aligned sequence length such as 2925 known to
   select a less-tested path?
2. In #36 you explained that the kernel performs its internal calculations in
   FP32 while importing/exporting FP16. Given finite inputs here, is there a
   known Volta tile, softmax, or reduction path that can still return NaNs? Are
   there required input-magnitude limits?
3. Could setting `softmax_scale=1/sqrt(128)` explicitly, or another existing
   option, select a safer kernel path?
4. Do you think this combination can be fixed with a Volta-specific stable
   accumulation/tile path? If it is unsupported, could the backend reject it
   explicitly instead of returning NaNs, so callers can fall back to SDPA?
5. Would a captured finite Q/K/V tensor set plus a Torch-SDPA reference output
   be sufficient for a standalone kernel reproducer? I can provide that data
   and test a candidate revision on the V100.

## Diagnostic code

- Probe and recorded result:
  https://github.com/buxinzi2233/MonadForge/blob/main/bench/v100_flash/README.md
- Finite-check implementation and isolation commit:
  https://github.com/buxinzi2233/MonadForge/commit/45221697987778b75f54e194f68d4568ea4f9d71

The host-side checks can be run with:

```bash
python -m bench.v100_flash.run_probe \
  --attn_mode torch --device cuda --steps 5

python -m bench.v100_flash.run_probe \
  --attn_mode flash --stability hybrid --debug_finite \
  --device cuda --steps 5

python -m bench.v100_flash.run_probe \
  --attn_mode flash --stability safe --debug_finite \
  --device cuda --steps 5
```

I would appreciate your view on whether this is a supported-shape bug and
whether a kernel-side fix is possible. I can provide additional captures or
run a focused test matrix if needed.

Thank you.

**Comment by ai-bond** [comment]
I don't really want to respond to a message generated by claw's bot...

> Thank you for maintaining FlashAttention for Volta. I am testing the public FA2-compatible API as the attention backend for Anima  iT training on a Tesla V100. 

It was really u or your bot )))

> I found a numerical failure where finite FP16 Q/K/V tensors produce a non-finite dense self-attention output on the first training step.

Are u sure really about ?
> it is therefore related to, but materially different from, #36, where the integration layer remained a possible cause.

So your bot "runs tests," finds something it thinks it can, and then asks me about support?
> Is SM70 + FP16 + head_dim=128 + seqlen=2925 supported by backend v26.06? In particular, is a non-tile-aligned sequence length such as 2925 known to select a less-tested path?

[flash_attn_v100/flash_attn_interface.py](https://github.com/ai-bond/flash-attention-v100/blob/main/flash_attn_v100/flash_attn_interface.py) 
```
        pad_size = 0
        if head_size_og % 8 != 0:
            pad_size = 8 - head_size_og % 8
            q_ = torch.nn.functional.pad(q_, [0, pad_size])
            k_ = torch.nn.functional.pad(k_, [0, pad_size])
            v_ = torch.nn.functional.pad(v_, [0, pad_size])

        q_ = q_.contiguous()
        k_ = k_.contiguous()
        v_ = v_.contiguous()
```
Same behavior like in TriDao's repo.

**Comment by ai-bond** [comment]
Of course it can - any code at any deployment stages can have issues!
>In https://github.com/ai-bond/flash-attention-v100/issues/36 you explained that the kernel performs its internal calculations in
FP32 while importing/exporting FP16. Given finite inputs here, is there a known Volta tile, softmax, or reduction path that can still  eturn NaNs? Are there required input-magnitude limits?

Before i will answer - explain true reason of using such param type!
>Could setting softmax_scale=1/sqrt(128) explicitly, or another existing option, select a safer kernel path?

From the question - more clearly see who is asked! I mean real person or bot )
>Do you think this combination can be fixed with a Volta-specific stable accumulation/tile path? If it is unsupported, could the backend reject it explicitly instead of returning NaNs, so callers can fall back to SDPA?

**Comment by buxinzi2233** [comment]
The attempt to train the Anima DIT architecture model using flash that supports v100 was some time ago. It was temporarily shelved at the time due to its inability to be used, and this issue is a compilation of documents from the repository that I had GPT5.6SOL create.
The initial problem was that training the Anima DIT architecture model with LoRa resulted in NaN errors with fp16 accuracy. The trainer's authors eventually resolved this by creating a solution using a mixed fp16/32 precision approach. We then tried training with Flash, and I recall that it still resulted in NaN errors, although my memory might be inaccurate due to the passage of time.

**Comment by buxinzi2233** [comment]
## Follow-up: captured tensors, independent replay, and full-pipeline confirmation

I ran an additional provenance and reproducibility audit of the report above. The evidence is archived here in my fork:

<https://github.com/buxinzi2233/MonadForge/releases/tag/diagnostic-v100-20260730>

Direct files:

- [original `first_failure.pt`](https://github.com/buxinzi2233/MonadForge/releases/download/diagnostic-v100-20260730/first_failure.pt) (84,588,422 bytes)
- [original `first_failure.json`](https://github.com/buxinzi2233/MonadForge/releases/download/diagnostic-v100-20260730/first_failure.json)
- [raw/compat/compiled replay](https://github.com/buxinzi2233/MonadForge/releases/download/diagnostic-v100-20260730/replay.json)
- [complete 4112..4128 alignment sweep](https://github.com/buxinzi2233/MonadForge/releases/download/diagnostic-v100-20260730/sweep_mod16.json)
- [independent full-training capture](https://github.com/buxinzi2233/MonadForge/releases/download/diagnostic-v100-20260730/independent_first_failure.pt) (16,160,257 bytes)
- [audit report](https://github.com/buxinzi2233/MonadForge/releases/download/diagnostic-v100-20260730/flash-attention-v100-data-collection-report.md)

### Original capture

The first real Anima training forward failed at block 0 self-attention with:

```text
Q/K/V: (1, 4130, 16, 128), torch.float16, all finite
attention output: (1, 4130, 2048), torch.float16
NaN / +Inf / -Inf: 1,412,224 / 562 / 590
finite output range: [-914, 840]
sigma: 0.7394310832
sample: post_image_dataset/resized/03.png::anima-multires=944x1120
```

The exact local SHA-256 of the original tensor capture is:

```text
91f67505dd66914718a3de61d361c71a1d621fcda46f0fa0d43731d11f05fa0d
```

The source image is 944x1120. Its latent cache and its one safetensors text cache containing four caption variants were independently checked: every floating-point array is finite.

### Independent replay on the V100

I reran the capture through the following paths on the same Tesla V100:

```bash
.venv/bin/python -m bench.v100_flash.replay_capture \
  output/daemon/jobs/20260729-finite-capture/finite-capture/first_failure.pt \
  --output /tmp/replay.json
```

Raw Flash eager, compatibility-wrapper eager, and compatibility-wrapper compiled all produced the same non-finite mask and counts. Torch SDPA on the exact same Q/K/V was fully finite. The rerun JSON is byte-for-byte identical to the original replay report (SHA-256 `c9a4c3fbd59b0cf7bd5e102a6d51776d9c4d83d7c916bbbd8c7cd1d4f01584bc`).

I also reran every prefix length from 4112 through 4128. 4112 and 4128 (multiples of 16) were finite; every length 4113 through 4127 was non-finite in raw V100 FlashAttention, while SDPA was finite for all 17 lengths. The categorical flags and Flash NaN counts matched the previous sweep exactly.

### Independent full training run

To verify that this was not merely a hand-created replay tensor, I loaded the saved training configuration and executed a new complete Anima forward with a fresh capture directory:

```bash
.venv/bin/python train.py \
  --config_file output/logs/lora-iumu-对比实验-te_20260729-2354/lora-iumu-对比实验-te.snapshot.toml \
  --no-config-snapshot \
  --finite_capture_dir /tmp/monadforge-auth-audit/finite-capture \
  --progress_jsonl /tmp/monadforge-auth-audit/progress.jsonl
```

This run loaded the same Anima model, LoRA, V100 fork, and dataset path. Because the old snapshot did not record a deterministic seed, the first batch was a different bucket, but the real model path failed in the same boundary-sensitive operation:

```text
sample: post_image_dataset/resized/03.png::anima-multires=464x544
Q/K/V: (1, 986, 16, 128), torch.float16, all finite
attention output: (1, 986, 2048), torch.float16
NaN / +Inf / -Inf: 116,736 / 68 / 60
986 % 16 = 10
```

The independent capture SHA-256 is:

```text
074a5870d073ffcb9b4240ac047ec96502a132a922fa8946df7f26254cd4194b
```

### Limits of the claim

I am not claiming a third-party tamper-proof attestation: the original local artifacts were not signed externally, and the original run did not preserve a complete stdout log or explicit seed. The claim supported by the data is narrower and reproducible: finite Q/K/V inputs on this V100 produce non-finite dense FlashAttention output for these non-16-aligned lengths, while SDPA remains finite. The full audit report in the release lists all hashes, timestamps, cache checks, and limitations.

Could you please test the attached `first_failure.pt` with the public `flash_attn_func` API and advise whether the dense non-aligned tail path is supported in v26.06? A kernel-side guard or a documented SDPA fallback would prevent downstream training from discovering this as NaN loss.

**Comment by buxinzi2233** [comment]
The above is an actual error data collection performed with the assistance of GPT.

**Comment by ai-bond** [comment]
All ur answer is bot code generation ) But ok.....

Funny marketing story with anthropic incident like escape from vm .....
> this issue is a compilation of documents from the repository that I had GPT5.6SOL create.

Because ) 
> The attempt to train the Anima DIT architecture model using flash that supports v100 was some time ago.

I've tried to train my model with 1,6B params from scratch with this lib and with 2xV100 16G successfully done it with totally random data with fp16 accuracy and fietuned with LoRA (using alpaca's datasets).
> The initial problem was that training the Anima DIT architecture model with LoRa resulted in NaN errors with fp16 accuracy.

))))))))) where is this `code solution`?
> The trainer's authors eventually resolved this by creating a solution using a mixed fp16/32 precision approach.

)))))))))
> Raw Flash eager, compatibility-wrapper eager, and compatibility-wrapper compiled

May be this is fine reason to use SDPA?
> all produced the same non-finite mask and counts. Torch SDPA on the exact same Q/K/V was fully finite.

)))))))
> The categorical flags and Flash NaN counts matched the previous sweep exactly.

Because inspect with GPT5.6SOL how mma work and multiply tiles with less than 16x16x16 and give for whole community code that's will solve this issue withoud speed degradation! Find out how SDPA do this! Why u only ask question?! Spend a bit tokens to find out reason and share with us.
> I also reran every prefix length from 4112 through 4128. 4112 and 4128 (multiples of 16) were finite
> while SDPA was finite for all 17 lengths.

Nice one try ))) to `escape from sandbox` ))))
> Could you please test the attached first_failure.pt with the public flash_attn_func API and advise whether the dense non-aligned tail path is supported in v26.06?

**Comment by ai-bond** [comment]
And? Reason for openning this "issue"? Do you have to offer something or we can close ticket?

**Comment by buxinzi2233** [comment]
The main issue is that I couldn't train normally using Flash + FP16 before, so I wanted to understand the possible reasons.

After contacting the trainer author and reporting the problem, he provided a new set of test documents and pointed out that this might be due to a density tail defect in the existing wheel package. After building based on the latest source code in your repository, I was able to train normally at 896 and 1024 resolutions. However, strangely, after training at 1536 resolution for a period of time (with gradient checkpointing enabled), the trainer's NaN warning protection mechanism was triggered.

Currently, due to the trainer author's health condition, further investigation is not possible. Therefore, my answer to this question is "Perhaps?"

Perhaps there is some problem with the current trainer's design.

**Comment by buxinzi2233** [comment]
Thanks for your help

**Comment by ai-bond** [comment]
From time to time i saw a lot of conversations in different code community's about "should we accept AI generated content or patches". This "issue" by clawbot's github accounts is show how social engineering is work. It tried generate fake problem, ask run unknown code for tests - and at the end "Currently, due to the trainer author's health condition, further investigation is not possible. Therefore, my answer to this question is "Perhaps?"" )))))))

Be careful!
